plugins/reports/api/reports.js

```diff
@@ -577,15 +599,62 @@ var metricProps = {
         }
     };
 
+    reports.decryptUnsubscribeCode = function (data) {
+        try {
+            const reportConfig = plugins.getConfig("reports", null, true);
+            const key = reportConfig.secretKey;
+            const decipher = crypto.createDecipheriv('aes-256-ctr', key, Buffer.from(data.iv, 'hex'));
+            const decrpyted = Buffer.concat([decipher.update(Buffer.from(data.content, 'hex')), decipher.final()]);
+            const result = JSON.parse(decrpyted.toString());
+            return result;
+        } catch (e) {
+            log.e("decrypt unsubscribe code err", e);
+        }
+    }
+
+    reports.genUnsubscribeCode = function (report, email) {
+        try {
+            const reportConfig = plugins.getConfig("reports", null, true);
+ 
+            const iv = crypto.randomBytes(16);
+            const key = reportConfig.secretKey;
+            const cipher = crypto.createCipheriv('aes-256-ctr', key, iv);
+            const data = {
+                "reportID": report._id,
+                "email": email,
+            };
+
+            const encrypted = Buffer.concat([cipher.update(JSON.stringify(data)), cipher.final()]);
+            const result = {
+                iv: iv.toString('hex'),
+                content: encrypted.toString('hex')
+            };
+            return JSON.stringify(result);
+        } catch (e) {
+            console.log(e, "[report subscription encode]");
+        }
+    }
+
     reports.send = function(report, message, callback) {
         if (report.emails) {
             for (var i = 0; i < report.emails.length; i++) {
                 var msg = {
                     to: report.emails[i],
                     from: versionInfo.title,
                     subject: report.subject,
-                    html: message
+                    // if report contains customize message for each email address, use reports.messages[i]
+                    html: report.messages && report.messages[i] && report.messages[i].html || message,
                 };
+
+                if (report.messages && report.messages[i]) {
+                    msg.list = {
```

Comment:
And can we add that `List-Unsubscribe` header with link too? I know it won't work in all cases, but should be a nice addition, for future.
It will mean modifying mail.js to accept custom headers too.

Reply:
Yes, we already add  "List-Unsubscribe" header  to nodemail lib, same example here: https://nodemailer.com/message/list-headers/

Reply:
Where do we add it?